contentcuration/contentcuration/utils/nodes.py

```diff
@@ -275,3 +286,189 @@ def generate_diff(updated_id, original_id):
+
+    def get_size(self):
+        size = self.cache_get(self.size_key)
+        return int(size) if size else size
+
+    def get_modified(self):
+        modified = self.cache_get(self.modified_key)
+        return isoparse(modified) if modified else modified
+
+    def set_size(self, size):
+        return self.cache_set(self.size_key, size)
+
+    def set_modified(self, modified):
+        return self.cache_set(self.modified_key, modified.isoformat() if isinstance(modified, datetime) else modified)
+
+
+class ResourceSizeHelper:
+    """
+    Helper class for calculating resource size
+    """
+    def __init__(self, node):
+        """
+        :param node: The contentnode with which to determine resource size
+        :type node: ContentNode
+        """
+        self.node = node
+
+    @property
+    def queryset(self):
+        """
+        :rtype: QuerySet
+        """
+        qs = self.node.get_descendants(include_self=True)
+
+        if self.node.is_root_node():
+            # since root node, remove unneeded filtering
+            qs = ContentNode.objects.filter(tree_id=self.node.tree_id)
+        # else if it's a leaf node, simplification handled by `get_descendants`
+
+        return File.objects.filter(contentnode__in=qs.filter(complete=True))
+
+    def get_size(self):
+        """
+        Calculates the size of the resource and it's descendants
+
+        SQL:
+            SELECT SUM(file_size)
+            FROM (
+                SELECT DISTINCT
+                    checksum,
+                    file_size
+                FROM contentcuration_file
+                WHERE contentnode_id IN ( SELECT id FROM contentcuration_contentnode WHERE ... )
+            ) subquery
+            ;
+
+        :return: An integer representing the resource size
+        """
+        sizes = self.queryset.values("checksum").distinct().aggregate(resource_size=Sum("file_size"))
+        return sizes['resource_size']
+
+    def modified_since(self, compare_datetime):
+        """
+        Determines if resources have been modified since ${compare_datetime}
+
+        SQL:
+            SELECT BOOL_OR(modified > ${compare_datetime})
+            FROM (
+                SELECT
+                    modified_at
+                FROM contentcuration_file
+                WHERE contentnode_id IN ( SELECT id FROM contentcuration_contentnode WHERE ... )
+            ) subquery
+            ;
+
+        :
+        :param compare_datetime: The datetime with which to compare.
+        :return: A boolean indicating whether or not resources have been modified since the datetime
+        """
+        compare_datetime = compare_datetime.isoformat() if isinstance(compare_datetime, datetime) else compare_datetime
+        result = self.queryset.aggregate(
+            modified_since=BoolOr(CombinedExpression(F('modified'), '>', Value(compare_datetime)))
+        )
+        return result['modified_since']
+
+
+STALE_MAX_CALCULATION_SIZE = 5000
```

Comment:
This should be conservative threshold for synchronous calculation. I added logging below to track the calculation time in case we need to adjust.